### PR TITLE
#251 & #461 add change outcome on screening

### LINKED ISSUE #251: Configure action for screening to change outcome to 'rejected'
(for review testing template)

### LINKED ISSUE #461: Update incrementStage and generateReviewAssignments to only run if application outcome is still PENDING - so it doesn't run after the outcome changes

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #251 
Fixes #461 

### Linked PR
- Base back-end PR #447 

### Changes
- Add check for application outcome is still PENDING before incrementing stage
- Create specific `changeOutcome` action in the Drug Registration template to move the outcome to REJECT in case the Screener (Stage: 1) set a review decision as NON_CONFORM.

### Problem
- Although the sequence number of `changeOutcome` specific to Drug Registration is defined as **10** and I changed the other core_actions to start from **50** the core_actions are running first, therefor the stage is getting changed before the changeOutcome action runs, the result is that the application moves to the next stage when it should have stayed in Screening.

### CODE REVIEW

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Just a couple of quick comments.

Also -- I'm having trouble testing it with front-end. I think you need to update this branch from `master` (and your other PR), as the current front-end master is inconsistent with this branch and doesn't run. :(

>     * Although the sequence number of `changeOutcome` specific to Drug Registration is defined as **10** and I changed the other core_actions to start from **50** the core_actions are running first, therefor the stage is getting changed before the changeOutcome action runs, the result is that the application moves to the next stage when it should have stayed in Screening.

Be good to get to the bottom of this, but haven't been able to test it yet. Are you sure they're running out of sequence? (i.e. have you checked the "action_queue" logs?) I've never seen sequential actions run out of order before, I think it's more likely that the incorrect Outcome data is getting to the 'incrementStage' action. Would be keen to investigate further...

**Comment by nmadruga** [comment]
> Also -- I'm having trouble testing it with front-end. I think you need to update this branch from `master` (and your other PR), as the current front-end master is inconsistent with this branch and doesn't run. :(

Will do this now. Thanks!

**Comment by nmadruga** [comment]
> Also -- I'm having trouble testing it with front-end. I think you need to update this branch from `master` (and your other PR), as the current front-end master is inconsistent with this branch and doesn't run. :(

Hi @CarlosNZ I have updated PR #460 with `master` and this one also. You should be ok with running this back-end with just master front-end (after running `yarn generate` to get latest types)

**Comment by nmadruga** [comment]
> Are you sure they're running out of sequence? (i.e. have you checked the "action_queue" logs?) I've never seen sequential actions run out of order before, I think it's more likely that the incorrect Outcome data is getting to the 'incrementStage' action. Would be keen to investigate further...

Can we have a look on this together after break please?

**Comment by nmadruga** [comment]
> Also -- I notice you didn't make any changes to `generateReviewAssignments` as described in #461 -- is this to be done in another PR?

Thanks! I was thinking, this is actually needed only to prevent **newly** added Reviewers (with permission) to not have reviewAssignments created, since it would run the action on the same stage/level and no new reviewAssignments would be generated unless a new Reviewer exists AND don't have reviewAssignments created yet. I'll just add this clause anyway to the `generateReviewAssignments` and others that might need to check for that (since they run ON_REVIEW_SUBMIT after the `changeOutcome` is done)

File: plugins/action_increment_stage/src/incrementStage.ts [hunk 1]
```diff
@@ -7,12 +11,19 @@ async function incrementStage({
...
     error_log: 'unknown error',
   }
   console.log(`Incrementing the Stage for Application ${applicationId}...`)
 
+  if (outcome != ApplicationOutcome.Pending) {
+    returnObject.status = ActionQueueStatus.Fail
+    returnObject.error_log = 'Outcome not pending'
+    return returnObject
+  }
+
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply plugins/action_increment_stage/src/incrementStage.ts:26]: Ok. Good point. Changing this now

**Comment by nmadruga** [comment]
> > Also -- I notice you didn't make any changes to `generateReviewAssignments` as described in #461 -- is this to be done in another PR?
> 
> Thanks! I was thinking, this is actually needed only to prevent **newly** added Reviewers (with permission) to not have reviewAssignments created, since it would run the action on the same stage/level and no new reviewAssignments would be generated unless a new Reviewer exists AND don't have reviewAssignments created yet. I'll just add this clause anyway to the `generateReviewAssignments` and others that might need to check for that (since they run ON_REVIEW_SUBMIT after the `changeOutcome` is done)

Ok. As discussed in person. No need to change `generateReviewAssignments` for now since it's not generating any new reviewAssignments for same stage and level unless a new reviewer is added which should be sorted in #429

File: plugins/action_increment_stage/src/incrementStage.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply plugins/action_increment_stage/src/incrementStage.ts]: Done 1eda225

**Comment by nmadruga** [comment]
@CarlosNZ is this one approved?

**Comment by CarlosNZ** [comment]
> @CarlosNZ is this one approved?

Yeah, I guess -- I assume all the problems are now fixed?

**Comment by nmadruga** [comment]
> > @CarlosNZ is this one approved?
> 
> Yeah, I guess -- I assume all the problems are now fixed?

Yes problem is fixed:
- In dev templates **Drug registration** and **Test Review** now setting Screening to NON CONFORM will change outcome of application to REJECTED
- After that action `incrementStage` will check that `outcome` is =! to PENDING and won't increment the stage
-

**Review by CarlosNZ — APPROVED** [review]

**Pull request merged**